src/useq/_mda_sequence.py

```diff
@@ -344,51 +418,55 @@ def iter_sequence(sequence: MDASequence) -> Iterator[MDAEvent]:
     MDAEvent
         Each event in the MDA sequence.
     """
-    order = sequence.used_axes
-
-    event_iterator = (enumerate(sequence.iter_axis(ax)) for ax in order)
-    for global_index, item in enumerate(product(*event_iterator)):
+    global_index = 0
+    event_iterator = (enumerate(sequence.iter_axis(ax)) for ax in sequence.used_axes)
+    for item in product(*event_iterator):
         if not item:  # the case with no events
             continue  # pragma: no cover
 
-        _ev = dict(zip(order, item))
-        index = {k: _ev[k][0] for k in INDICES if k in _ev}
-
-        position: Optional[Position] = _ev[POSITION][1] if POSITION in _ev else None
-        channel: Optional[Channel] = _ev[CHANNEL][1] if CHANNEL in _ev else None
-        time: Optional[int] = _ev[TIME][1] if TIME in _ev else None
-        grid: Optional[GridPosition] = _ev[GRID][1] if GRID in _ev else None
+        _ev, index = sequence._get_event_and_index(item)
+        position, channel, time, grid = sequence._get_axis_info(_ev)
 
         # skip channels
         if channel and TIME in index and index[TIME] % channel.acquire_every:
             continue
-
-        if grid:
-            x_pos: Optional[float] = grid.x
-            y_pos: Optional[float] = grid.y
-            if grid.is_relative:
-                px = getattr(position, "x", 0) or 0
-                py = getattr(position, "y", 0) or 0
-                x_pos = x_pos + px if x_pos is not None else None
-                y_pos = y_pos + py if y_pos is not None else None
-        else:
-            x_pos = getattr(position, "x", None)
-            y_pos = getattr(position, "y", None)
+        # skip if in position sequence
+        if position and position.sequence and CHANNEL in index and index[CHANNEL] != 0:
+            continue
+        if position and position.sequence and GRID in index and index[GRID] != 0:
+            continue
+        if position and position.sequence and Z in index and index[Z] != 0:
+            continue
 
         try:
-            z_pos = (
-                sequence._combine_z(_ev[Z][1], index[Z], channel, position)
-                if Z in _ev
-                else position.z
-                if position
-                else None
-            )
+            z_pos = sequence._get_z(_ev, index, position, channel)
         except sequence._SkipFrame:
             continue
 
+        # position sequence
+        if position and position.sequence:
+            p_sequence = position.sequence
+
+            p_event_iterator = (
+                enumerate(p_sequence.iter_axis(ax)) for ax in p_sequence.used_axes
+            )
+            for p_item in product(*p_event_iterator):
+                yield _iter_position_sequence(
+                    sequence, p_sequence, index, p_item, position, global_index
+                )
+                global_index += 1
+            continue
```

Comment:
notice how `_iter_position_sequence` is pretty much the same as `iter_sequence`?  use recursion! :)

```suggestion
        if position and position.sequence:
            for sub_event in iter_sequence(position.sequence):
                subz = (z_pos or 0) + sub_event.z_pos if sub_event.z_pos else z_pos
                subx = (x_pos or 0) + sub_event.x_pos if sub_event.x_pos else x_pos
                suby = (y_pos or 0) + sub_event.y_pos if sub_event.y_pos else y_pos
                yield sub_event.copy(
                    update={
                        "global_index": global_index,
                        "index": {**index, **sub_event.index},
                        "x_pos": subx,
                        "y_pos": suby,
                        "z_pos": subz,
                        "sequence": sequence,
                    }
                )
                global_index += 1
            continue
```

and then delete the `_iter_position_sequence` function below.

then actually, that whole bit about `subz`, `subx` etc... could be a method on the MDAEvent class to essentially "shift" its position a bit, returning a new MDA event object